Try to fix formatting failures on Windows (issue #2810)

dfyz opened the pull request:
When `newline_style` is set to `Windows`, an empty line inside of a macro results in `\r` being passed to the `fold()` in `MacroBranch::rewrite()`. `\r` is technically _not_ an empty string, so we try to indent it, leaving trailing whitespaces behind, even thought that was not intended (as far as I can see).

There is a handy `is_empty_line()` function in the same file. This PR replaces the `!l.is_empty()` check with calling this function (since trying to indent any whitespace-only string will probably result in trailing whitespaces), but I'm not really sure if this is the best way to fix the root problem. Any feedback is appreciated.

otavio commented:
Hello! thank you for the pull request. It seems both source and target files are the same on the test, in this case only the target file is required.

dfyz commented:
Thank you for the wonderful tool!

The files are actually subtly different (the source side has trailing whitespaces in the macro, the target side does not):
```
$ diff -u tests/{source,target}/issue-2810.rs 
--- tests/source/issue-2810.rs	2018-10-27 05:20:56.360164064 +0300
+++ tests/target/issue-2810.rs	2018-10-27 05:20:56.360164064 +0300
@@ -6,7 +6,7 @@
         pub struct $result {
             success: bool,
         }
-        
+
         impl $result {
             pub fn foo(self) {}
         }
```
I just copied the test case verbatim from the original issue (just to be clear, this is not my code). Since the bug doesn't depend on the presence of trailing whitespaces, I can remove them from the source side to make it less confusing. Does this sound good?

I'd also appreciate some help with failing test cases on Travis. The last three builds ([#4598](https://travis-ci.org/rust-lang-nursery/rustfmt/builds/445969561), [#4600](https://travis-ci.org/rust-lang-nursery/rustfmt/builds/446573007), [#4602](https://travis-ci.org/rust-lang-nursery/rustfmt/builds/446937668)) all fail different tests. Is this expected? How do I find out which regressions were introduced by me?

otavio commented:
> Thank you for the wonderful tool!

I am just starting to contribute, like you. Please be welcome aboard :-)

> The files are actually subtly different (the source side has trailing whitespaces in the macro, the target side does not)

> I just copied the test case verbatim from the original issue (just to be clear, this is not my code). Since the bug doesn't depend on the presence of trailing whitespaces, I can remove them from the source side to make it less confusing. Does this sound good?

It makes sense; the smaller the test case, more objective it is. So please amend your commit doing this change :+1: 

> I'd also appreciate some help with failing test cases on Travis. The last three builds ([#4598](https://travis-ci.org/rust-lang-nursery/rustfmt/builds/445969561), [#4600](https://travis-ci.org/rust-lang-nursery/rustfmt/builds/446573007), [#4602](https://travis-ci.org/rust-lang-nursery/rustfmt/builds/446937668)) all fail different tests. Is this expected? How do I find out which regressions were introduced by me?

In fact, it is on the integration test script. Take a look at https://github.com/rust-lang-nursery/rustfmt/blob/master/ci/integration.sh that is the script called for it.

In summary, it uses `cargo` source as a test case, however, it seems the code has not been formatted using the last release yet, hence the failure.

dfyz commented:
I simplified the test and rewrote the commit message, adding `Fixes: #2810` along the way.

> it seems the code has not been formatted using the last release yet, hence the failure

Unfortunately, I don't think this is the reason for Travis failures. The comments in the testing script say the following:
```
# Checks that:
#
# * `cargo fmt --all` succeeds without any warnings or errors
# * `cargo fmt --all -- --check` after formatting returns success
# * `cargo test -all` still passes (formatting did not break the build)
```
I take it that this sequence of commands should succeed on any Rust project, no matter what version of rustfmt was originally used to format it.

My understanding is that there is currently a non-deterministic behavior in the `master` version of rustfmt. I can reproduce it by checking out locally any of the repos used in the integration tests (say, `rand`), installing rustfmt from `master` with `cargo +nightly install --force` and then running `cargo fmt --all && cargo fmt --all -- --check` in the checked out `rand/` directory. The nightly version of rustfmt always exits successfully whereas the `master` version fails on the `--check` call approximately 9 times out of 10 (each time with a _different_ diff). I suppose the integration tests on Travis fail because of this.

`git bisect` says the commit de0b661bfa7646b1c871b3b10de2672cd911e6d0 introduced non-determinism. I presume this is going to be fixed in #3143, although I'm not 100% sure.

otavio commented:
I think you are right, good catch.

Please take a look at the #3143 and check if it fixes the regression for you.

otavio commented:
> `git bisect` says the commit [de0b661](https://github.com/rust-lang-nursery/rustfmt/commit/de0b661bfa7646b1c871b3b10de2672cd911e6d0) introduced non-determinism. I presume this is going to be fixed in #3143, although I'm not 100% sure.

Just did the check locally and #3143  does fix the issue.

dfyz commented:
Good news, thanks for the fix! I'll re-run the integration tests on Travis when #3144 is merged into the `master` branch.

otavio commented:
I reported https://github.com/rust-lang-nursery/crater/issues/358 as `crater` is failing. I will also open a PR to change it to allow failures of it.

dfyz commented:
This is awesome, thank you again!

nrc commented:
Thank you!
